### PR TITLE
Improve query logging

### PR DESCRIPTION
**Pull request by lachlan1m**
For each graphql request a single json blob will be logged containing any useful information.

### CODE REVIEW

**Comment by lachlan1m** [comment]
#695 #633

File: packages/query/src/graphql/plugins/logGraphql.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import {ApolloServerPlugin} from 'apollo-server-plugin-base';
+import {GraphQLRequestListener} from 'apollo-server-plugin-base/src/index';
+import {GraphQLRequestContext} from 'apollo-server-types';
+import {getLogger} from '../../utils/logger';
+
+const logger = getLogger('graphql');
+
+const getSizeInBytes = (obj: Record<string, unknown>) => {
+  let str: string;
+  if (typeof obj === 'string') {
```
Comment (zhex) [inline packages/query/src/graphql/plugins/logGraphql.ts:14]: the obj is Record type, is it still possible to be string?

Comment (lachlan1m) [reply packages/query/src/graphql/plugins/logGraphql.ts:14]: made a mistake here, forgot to update.

File: packages/query/src/graphql/plugins/logGraphql.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                //Warn as no metadata field should be null
+                logger.warn(`${info.parentType.name}.${info.fieldName} returned null`);
+              }
+            };
+          },
+        };
+      },
+      async willSendResponse({response}) {
+        if (!response.errors) {
+          logger.debug(logSizeInKilobytes('response size:', response.data));
```
Comment (ianhe8x) [inline packages/query/src/graphql/plugins/logGraphql.ts:60]: I don' think we want multiple lines for one request. Can you put everything about one request in one line?
and maybe format it as json so backend can more easily consume it. @zhex which formate do you prefer?

Comment (zhex) [reply packages/query/src/graphql/plugins/logGraphql.ts:60]: in the production, the debug level will be turned off, so it is good to add response data in the existing info log payload

Comment (lachlan1m) [reply packages/query/src/graphql/plugins/logGraphql.ts]: I have 1 line for graphql payload (doesn't convert nicely to json) and 1 json object for response data.

File: packages/query/src/graphql/plugins/logGraphqlPlugin.ts [hunk 3]
```diff
@@ -6,26 +6,26 @@ import {ApolloServerPlugin} from 'apollo-server-plugin-base';
...
-  const str = JSON.stringify(obj);
-  return new TextEncoder().encode(str).length;
-};
+const logger = getLogger('graphql');
 
 const getSizeInKilobytes = (obj: Record<string, unknown>) => {
-  const bytes = getSizeInBytes(obj);
-  const kb = (bytes / 1000).toFixed(2);
-  return kb;
+  const str = JSON.stringify(obj);
```
[the remark this answers is not in the document]
Comment (ianhe8x) [reply packages/query/src/graphql/plugins/logGraphqlPlugin.ts:20]: I think the better place to get the response size is from express response object. Though the down side is you will have to refactor the code.

File: packages/query/src/graphql/plugins/logGraphqlPlugin.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    if (httpHeaders) {
+      payload['x-forwarded-for'] = httpHeaders['x-forwarded-for'] ?? '127.0.0.1';
+      payload['user-agent'] = httpHeaders['user-agent'] ?? 'unknown';
+      payload.referer = httpHeaders.referer ?? 'unknown';
+    }
+
+    return {
+      async willSendResponse({response}) {
+        payload['response-time'] = `${Math.round(performance.now() - start)}`;
+        payload['response-payload-size'] = getSizeInKilobytes(response.data);
```
Comment (ianhe8x) [inline packages/query/src/graphql/plugins/logGraphqlPlugin.ts:56]: try this `response.http.headers.get('Content-Length')` though i'm not sure if it is set in this stage.

File: packages/query/src/graphql/plugins/logGraphqlPlugin.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ianhe8x) [reply packages/query/src/graphql/plugins/logGraphqlPlugin.ts:20]: I saw in line 67 the response is stringified again, if `JSON.stringify` is inevitable, at least reduce it to once.

File: packages/query/src/graphql/plugins/logGraphqlPlugin.ts (hunk 4 above)
Comment (lachlan1m) [reply packages/query/src/graphql/plugins/logGraphqlPlugin.ts:56]: My solution so far is to use response.http.headers.set(graphqlData), and then use ExpressPinoLogger with custom serializers.

**Pull request closed without merging**